### PR TITLE
8258444: Clean up specifications of java.io.Reader.read(char[],int,int) in subclass overrides

### PR DESCRIPTION
**Pull request by bplb**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258444](https://bugs.openjdk.java.net/browse/JDK-8258444): Clean up specifications of java.io.Reader.read(char[],int,int) in subclass overrides

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2680/head:pull/2680`
`$ git checkout pull/2680`

### CODE REVIEW

**Comment by bplb** [comment]
This change proposes to use {@inheritDoc} to improve consistency of the specification of `Reader.read(char[],in,in)` in `Reader` and its subclasses. In the course of this, some bounds checks are replaced with `Objects.checkFromIndexSize()`.

File: src/java.base/share/classes/java/io/InputStreamReader.java [hunk 1]
```diff
@@ -162,20 +162,10 @@ public int read() throws IOException {
...
-     * @return     The number of characters read, or -1 if the end of the
-     *             stream has been reached
-     *
-     * @throws     IOException  If an I/O error occurs
-     * @throws     IndexOutOfBoundsException {@inheritDoc}
+     * {@inheritDoc}
      */
-    public int read(char cbuf[], int offset, int length) throws IOException {
-        return sd.read(cbuf, offset, length);
+    public int read(char[] cbuf, int off, int len) throws IOException {
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/InputStreamReader.java:167]: IOOBE is unchecked, are you sure it gets inherited into the sub-class here?

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/io/InputStreamReader.java (hunk 1 above)
Comment (bplb) [reply src/java.base/share/classes/java/io/InputStreamReader.java:167]: It does not due to https://bugs.openjdk.java.net/browse/JDK-8157677.

Comment (AlanBateman) [reply src/java.base/share/classes/java/io/InputStreamReader.java:167]: The long standard behavior is that javadoc doesn't copy the  throws of unchecked exceptions, instead we've had to declare the throws and use inheritDoc. I assume you'll need to do that here.

Comment (bplb) [reply src/java.base/share/classes/java/io/InputStreamReader.java:167]: It does not and there is a javadoc issue already on file about this.

**Review by AlanBateman — APPROVED** [review]

**Comment by bplb** [comment]
CSR filed: https://bugs.openjdk.java.net/browse/JDK-8262262.

**Review by RogerRiggs — APPROVED** [review]
Nice cleanup.

File: src/java.base/share/classes/java/io/PushbackReader.java [hunk 2]
```diff
@@ -92,28 +93,14 @@ public int read() throws IOException {
...
-     *
-     * @return     The number of characters read, or -1 if the end of the
-     *             stream has been reached
-     *
-     * @throws     IOException  If an I/O error occurs
-     * @throws     IndexOutOfBoundsException {@inheritDoc}
+     * {@inheritDoc}
      */
-    public int read(char cbuf[], int off, int len) throws IOException {
+    public int read(char[] cbuf, int off, int len) throws IOException {
```
Comment (dfuch) [inline src/java.base/share/classes/java/io/PushbackReader.java:98]: Shouldn't you add

```
 * @throws IndexOutOfBoundException {@inheritDoc}
 * @throws IOException {@inheritDoc}
```

here as well? IIRC the global {@inheritDoc} will not add them.

Comment (bplb) [reply src/java.base/share/classes/java/io/PushbackReader.java:98]: No. In this case the specification no longer appears in the main method summary but rather under `Methods declared in class java.io.FilterReader` which has a full spec.

**Comment by bplb** [comment]
@AlanBateman, @dfuch, @RogerRiggs thanks for the reviews here and @jddarcy for approving the CSR!

/integrate

**Pull request closed without merging**